linkding: init at 1.36.0

l0b0 opened the pull request:
## Description of changes

From the [project](https://github.com/sissbruecker/linkding):

>  Self-hosted bookmark manager that is designed be to be minimal, fast, and easy to set up using Docker.

Closes #341665.

FYI @sissbruecker

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

l0b0 committed:
linkding: init at 1.36.0

Helps with #341665.

Co-authored-by: Pol Dellaiera <pol.dellaiera@protonmail.com>
Co-authored-by: Arne Keller <2012gdwu+github@posteo.de>
Co-authored-by: Gaétan Lepage <33058747+GaetanLepage@users.noreply.github.com>

trzpiot commented:
I can take over the review for aarch64-darwin as soon as the PR is ready. 🙂

l0b0 commented:
Some open issues at this point:

- [ ] I copied all the [dev dependencies](https://github.com/sissbruecker/linkding/blob/master/requirements.dev.in) into `nativeCheckInputs`. Trim down to only the ones needed for the build to pass.
- [ ] I copied all the [non-dev dependencies](https://github.com/sissbruecker/linkding/blob/621aedd8eb709a4f9cd3e59a9262ca002890622c/requirements.in) I could find in nixpkgs into `nativeBuildInputs`, but the build seems to pass even without some of them (commented out django-registration and uWSGI for now). Figure out whether they are actually needed, and remove any which are not needed.

drupol commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,86 @@
+  src = fetchFromGitHub {
+    owner = "sissbruecker";
+    repo = "linkding";
+    rev = "v${version}";
```
```suggestion
    rev = "refs/tags/v${version}";
```

drupol commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,86 @@
+    updateScript = nix-update-script { };
+  };
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

drupol commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,86 @@
+  meta = with lib; {
+    description = "Self-hosted bookmark manager";
+    homepage = "https://github.com/sissbruecker/linkding";
+    license = licenses.mit;
```
```suggestion
    license = lib.licenses.mit;
```

drupol commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,86 @@
+    description = "Self-hosted bookmark manager";
+    homepage = "https://github.com/sissbruecker/linkding";
+    license = licenses.mit;
+    maintainers = with maintainers; [ l0b0 ];
```
```suggestion
    maintainers = with lib.maintainers; [ l0b0 ];
```

drupol commented on nixos/modules/services/web-apps/linkding.nix:
```diff
@@ -0,0 +1,41 @@
+        LD_HOST_PORT = builtins.toString config.services.linkding.port;
+      };
+
+      serviceConfig.ExecStart = "${config.services.linkding.package}/bin/linkding";
```
use `lib.getExe config.services.linkding.package`

drupol reviewed (changes requested):
Hi!

The diff LGTM, except a few nits.

The PR must contain 2 commits ad minima:
 - 1 commit to add the package: `linkding: init at 1.36.0`
 - 1 commit to add the NixOS service: `nixos/linkding: init` (please verify in the contributor guide documentation how to format commit log messages correctly, I may be wrong here)
 
Since you're adding a new NixOS service, a release note entry must be added too.

l0b0 committed:
nixos/linkding: init

Closes #341665.

Includes all relevant options
<https://linkding.link/options/#list-of-options>, leaving any options
with a `null`/`[]` default unset if not explicitly set to something
else.

Flips the semantics of some options to avoid double negatives in the
configuration.

Co-Authored-By: Pol Dellaiera <pol.dellaiera@protonmail.com>

l0b0 commented:
> The PR must contain 2 commits ad minima:
> 
> * 1 commit to add the package: `linkding: init at 1.36.0`
> * 1 commit to add the NixOS service: `nixos/linkding: init` (please verify in the contributor guide documentation how to format commit log messages correctly, I may be wrong here)
> 
> Since you're adding a new NixOS service, a release note entry must be added too.

Thank you! All done.

drupol reviewed (approved):
LGTM

FliegendeWurst commented on nixos/doc/manual/release-notes/rl-2411.section.md:
```diff
@@ -108,6 +108,8 @@
 
 - [Eintopf](https://eintopf.info), community event and calendar web application. Available as [services.eintopf](options.html#opt-services.eintopf).
 
+- [linkding](https://linkding.link/), self-hosted bookmark manager. Available as [services.linkding](options.html#opt-services.linkding).
```
Should go in 25.05, unless you would prefer to backport.

FliegendeWurst commented on nixos/modules/services/web-apps/linkding.nix:
```diff
@@ -0,0 +1,41 @@
+
+    port = lib.mkOption {
+      default = 9090;
+      description = "Port on which linkding will listen for connections.";
```
I think a host option would also be useful. Values could for example be "127.0.0.1" (only available on localhost), "0.0.0.0" (available elsewhere)

FliegendeWurst commented on nixos/tests/web-apps/linkding/defaults.nix:
```diff
@@ -0,0 +1,19 @@
+    };
+
+    testScript = ''
+      ${nodeName}.wait_for_unit("linkding.service")
```
Would also be good to request "/" or some other path, just to make sure it responds to requests too

FliegendeWurst commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,86 @@
+  nativeBuildInputs = [
+    nodejs
+    npmHooks.npmConfigHook
+    python3Packages.beautifulsoup4
+    python3Packages.bleach-allowlist
+    python3Packages.django-widget-tweaks
+    python3Packages.djangorestframework
+    python3Packages.huey
+    python3Packages.markdown
+    python3Packages.mozilla-django-oidc
+    python3Packages.python-dateutil
+    python3Packages.waybackpy
```
I think these also go in `build-system`

FliegendeWurst replied:
Or actually in `dependencies`, they don't look build-related.

FliegendeWurst commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,86 @@
+    python3Packages.playwright
+  ];
+
+  checkPhase = ''
+    python manage.py test bookmarks.tests
+
+    export PLAYWRIGHT_BROWSERS_PATH=${pkgs.playwright-driver.browsers}
+    export PLAYWRIGHT_SKIP_VALIDATE_HOST_REQUIREMENTS=true
+    python manage.py test bookmarks.e2e --pattern="e2e_test_*.py"
+  '';
```
```suggestion
  checkPhase = ''
    runHook preCheck
    python manage.py test bookmarks.tests

    export PLAYWRIGHT_BROWSERS_PATH=${pkgs.playwright-driver.browsers}
    export PLAYWRIGHT_SKIP_VALIDATE_HOST_REQUIREMENTS=true
    python manage.py test bookmarks.e2e --pattern="e2e_test_*.py"
    runHook postCheck
  '';
```

FliegendeWurst commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,86 @@
+
+  meta = {
+    description = "Self-hosted bookmark manager";
+    homepage = "https://github.com/sissbruecker/linkding";
```
```suggestion
    homepage = "https://github.com/sissbruecker/linkding";
    changelog = "https://github.com/sissbruecker/linkding/blob/master/CHANGELOG.md";
```
(or use a version-specific link if you prefer)

l0b0 commented:
Converted to draft since this is actually *much* harder than I originally thought. This is a Django application built with Docker-only deployments in mind, and will require at least twice as much code to actually work as intended.

l0b0 commented:
Update: Got a bit further; any other tests/improvements you'd like to see?

l0b0 commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,134 @@
+    python3Packages.confusable-homoglyphs
+    python3Packages.cryptography
+    python3Packages.django
+    # python3Packages.django-registration # Not in nixpkgs
```
I don't know whether this is actually needed for anything.

FliegendeWurst replied:
Yes, I think it is, if you enable ALLOW_REGISTRATION

https://github.com/sissbruecker/linkding/blob/d2e8a95e3cccd58d7d08be9692a5e6ed5945b2af/siteroot/urls.py#L82-L83

l0b0 committed:
python312Packages.django-registration: init at 5.1.0

Co-authored-by: Gaétan Lepage <33058747+GaetanLepage@users.noreply.github.com>

GaetanLepage reviewed (commented):
FYI, linkding 1.37.0 is out (it's ok to not bump it just yet).

l0b0 commented on pkgs/development/python-modules/django-registration/default.nix:
```diff
@@ -0,0 +1,56 @@
+  version = "5.1.0";
+  pyproject = true;
+
+  disabled = pythonOlder "3.9" || pythonAtLeast "3.13";
+
```
Would you be willing to explain why? Nixpkgs currently has Python 3.14, and it would probably be bad for anyone to try to use that version with this package.

l0b0 replied:
That's fair. I generally prefer keeping both so that it's easier to remember updating it. There's 563 matches for `pythonOlder "3.9"` in current master.

l0b0 replied:
OK, I'll just leave it for now then. Thanks for taking the time!

l0b0 commented on pkgs/development/python-modules/django-registration/default.nix:
```diff
@@ -0,0 +1,56 @@
+    nox
+  ];
+
+  check = false;
```
No idea why I did this originally; the tests are passing. Removed.

l0b0 commented on pkgs/development/python-modules/django-registration/default.nix:
```diff
@@ -0,0 +1,56 @@
+
+  disabled = pythonOlder "3.9" || pythonAtLeast "3.13";
+
+  src = fetchPypi {
```
I don't remember the reason anymore, but it seems to work with `fetchFromGitHub` :grin:. Thanks for the heads up!

l0b0 commented on pkgs/by-name/li/linkding/package.nix:
```diff
@@ -0,0 +1,134 @@
+{
+  fetchFromGitHub,
+  fetchNpmDeps,
+  lib,
+  nix-update-script,
+  nixosTests,
+  nodejs,
+  npmHooks,
+  pkgs,
+  playwright-driver,
+  python3Packages,
+  stdenvNoCC,
+  uwsgi,
```
I prefer to sort them alphabetically, to make it easier to find whether a specific entry is in the list, rather than by order of appearance, because I don't know how the latter helps anyone read the code. But I'll move `lib` and `stdenvNoCC`, and I'd be willing to listen to arguments for why order of appearance is useful :smile:.

FliegendeWurst commented:
Almost all tests fail with:
```
  - [pid=1621][err] Old Headless mode has been removed from the Chrome binary. Please use the new Headless mode (https://developer.chrome.com/docs/chromium/new-headless) or the chrome-headless-shell which is a standalone implementation of the old Headless mode (https://developer.chrome.com/blog/chrome-headless-shell).
 ```

supermarin commented:
@l0b0 need a hand pushing this one through? stumbled upon this one just now looking for linkding in nixos man. I rebased it in my fork https://github.com/supermarin/nixpkgs/commit/3ea635399e31cd0266cb60bc0f2b58b8f99b4b50 but not sure I can contribute the change here without opening a separate PR to nixpkgs. Feel free to reset on my tree and force push

l0b0 commented:
> @l0b0 need a hand pushing this one through? stumbled upon this one just now looking for linkding in nixos man. I rebased it in my fork [supermarin@3ea6353](https://github.com/supermarin/nixpkgs/commit/3ea635399e31cd0266cb60bc0f2b58b8f99b4b50) but not sure I can contribute the change here without opening a separate PR to nixpkgs. Feel free to reset on my tree and force push

Yes, please, I basically put this on indefinite pause.

l0b0 commented:
> @l0b0 no prob. want me to open a separate PR? don't think I can contribute to this one

Yeah, I don't think you can. Go for it, use as much as you like from this one.

fugidev commented:
> Sounds good I'll open a new one shortly. I saw there were a few releases since and I'll update it to 1.39. One thing that comes to my mind looking at this: does it make sense to duplicate every ENV setting in nix (might be long term maintenance burden) vs dynamically generate ENV from the module the fly? I see this much like how docker-compose based projects work

A freeform module would be better imo

supermarin commented:
ok so I spent a bit of time on this yesterday #404962, and I can't figure out why headless chromium is crashing on e2e tests. anyone has ideas?

pyrox0 commented:
#509346 supercedes this
